### ISSUE TITLE
Add support for visionOS targets

### ISSUE THREAD

**Issue by conradev** [rust-lang/compiler-team#642]
# Proposal

Apple recently released the first version of the [visionOS SDK](https://developer.apple.com/visionos/) for their Apple Vision Pro device. The device has standalone operating system and corresponding SDK called xrOS, much like the Apple Watch has watchOS or the Apple TV has tvOS.

Apple is using the platform names `xros` and `xrsimulator` for the Vision devices and Vision simulator, respectively. In this proposal, I want to add these new targets to Rust:

- `aarch64-apple-xros`
- `aarch64-apple-xros-sim`
- `x86_64-apple-xros-sim`

The corresponding LLVM targets are:

- `arm64-apple-xros`
- `arm64-apple-xros-simulator`
- `x86_64-apple-xros-simulator`

There is also `arm64e-apple-xros`, but afaik Rust doesn't support PAC yet on Apple OSes

# Mentors or Reviewers

- conradev

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [ ] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

# Comments

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

**Comment by Nilstrieb** [comment]
you need to fill out [all the questions](https://doc.rust-lang.org/rustc/target-tier-policy.html#adding-a-new-target) for new targets

**Comment by thomcc** [comment]
Isn't that just needed for the impl PR?

**Comment by thomcc** [comment]
I can volunteer for the implementation or the review. I was planning on doing the implementation for these anyway, as just did tvOS, but don't mind if someone else does this.

However, as discussed in Zulip there are some issues that prevent working on this in the near term:
- We're blocked by Apple upstreaming LLVM support for these -- doing it based on the LLVM from the preview SDK (similar to how we used to need to use the LLVM from the watchOS SDKs for those targets) is very painful.
- It is likely that these will not be the final target names.

**Comment by workingjubilee** [comment]
> Isn't that just needed for the impl PR?

An MCP is technically not required at all. Rather, what is actually important is the requirements of the [target tier policy] are addressed by the author and the compiler team reviews it. An MCP is simply a vessel for that review.

[target tier policy]: https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html

**Comment by coolbluewater** [comment]
Hi folks, what is the status of this? 

@thomcc 
>It is likely that these will not be the final target names.

I'm sure many developers, including myself, are eagerly waiting this support.
To unblock here are a couple of choices:
1) Adopt the proposed target names based on 'xros' in nightly and change them down the road if Apple changes them, or
2) Adopt the seemingly committed brand name of 'visionOS' (which is what appears in user-facing text e.g. in Xcode). 

Personally I prefer option #2 - it makes sense even if Apple continues to use the 'xros` name internally. We already have the precedent of a divergent naming convention in Rust for macOS ('darwin'), where both macOS and darwin seem reasonable choices.

Either way, hope that this issue progresses quick. It's problematic for a major OS to be impeded in this way, especially for simple Tier 3 support.

Thanks a lot folks!

**Comment by workingjubilee** [comment]
> Also, what would it take to get a private build of Rust working against visionOS? Two or three versions of LLVM is a long time! :)

There is a [dev guide chapter on how to build the compiler](https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html). The Rust build system's `config.toml` has keys for specifying what LLVM you are building against and how you want to link (`llvm-config` and `link-shared`, amongst others). I recommend asking on Zulip if you need additional help building the compiler.

**Comment by thomcc** [comment]
It's unclear how viable this is in practice since `llvm-config` isn't actually shipped with the XCode SDK, and it's not clear that everything else we need is shipped with it either.

My initial suspicion that we could use Apple's fork of LLVM (like we used to need to do for watchOS) might be wrong, since [it doesn't seem to include `xros`](https://github.com/search?q=repo%3Aapple%2Fllvm-project%20xros&type=code) [or `visionos`](https://github.com/search?q=repo%3Aapple%2Fllvm-project%20visionos&type=code), although I haven't looked carefully, perhaps it's in some branch.

If you get rustc to build against an LLVM that knows about visionos (or have questions), feel free to ping me and I can give some advice. Note that you'll be likely have to maintain a fork of rustc for a while to do this even if you get all that working, although perhaps we could merge some of the changes with the caveat that "for now you also need to build rustc against system LLVM").

It's also possible you could target the, err, target without direct LLVM support by building against a generic mach-o target and using various system tools to fix up the compilation artifacts, but it would likely be... quite painful.

**Comment by apiraino** [comment]
Closing this MCP on the concern that there is no LLVM support yet.

**Comment by lcapkovic** [comment]
Is this PR all that's needed to unblock this?
https://github.com/llvm/llvm-project/pull/77707

**Comment by agg23** [comment]
I spent a decent amount of time trying to get vrOS working in the interim. I'm currently blocked on trying to build `core` which is failing with `SIGTRAP`, which I can only imagine is somehow LLVM is broken from the very minor changes adding xrOS.

My Zulip conversation asking about this build issue:
https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Adding.20new.2C.20near.20duplicate.20target.2E.20Failing.20building.20.60libc.60

LLVM changes (for some reason Apple's changes appear to not be complete):
https://github.com/rust-lang/llvm-project/compare/rustc/17.0-2023-12-14...agg23:llvm-project:xrOS-target

`rust`:
https://github.com/rust-lang/rust/compare/master...agg23:rust:xrOS-target

`libc`:
https://github.com/rust-lang/libc/compare/main...agg23:libc:main

If anyone has any feedback, I'd love to hear it. I really want to get this working, and will help with getting this merged eventually (if that's desired), but for now I just want us to be able to target visionOS at launch.

**Comment by agg23** [comment]
To update: Apple's LLVM change is actually far from complete, and some core pieces of information are missing (such as IDs for new load commands in MachO). My build issues were solely due to LLVM being unable to actually build for the target due to these issues. I don't think that any meaningful progress will be made here until Apple graces us with more information.

**Comment by phatblat** [comment]
@agg23 It looks like some MachO stuff was added in llvm/llvm-project#78373. Does this include the load commands you were referring to?

**Comment by phatblat** [comment]
FYI, I don't think the `x86_64-apple-xros-simulator` target will be valid because it's called out very clearly in the Xcode 15.2 [release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_2-release-notes): 

> Developing for visionOS requires a Mac with Apple silicon. (114799042)

**Comment by agg23** [comment]
No, I'm monitoring the `llvm-project` repo for any PRs submitted by Apple, and have been incrementally testing builds to see if that has the missing piece. What I am expecting is a `LC_VERSION_MIN_XROS` load command key (https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/MachO.def#L74) though I'm not sure that it's actually required (I don't really know what I'm doing here).

Apple appears to be making targeted changes that don't add xrOS all throughout LLVM, but just to very particular parts, and they're PRing incomplete segments of code (which I personally rather dislike), so it's hard to tell when it is usable.

----

I may have a working version of LLVM, but I don't know how to test outside of `rustc`. When attempting to build the stage 2 compiler libraries, which is the first time the compiler is used to produce native binaries (so `aarch64-apple-macos` building `aarch64-apple-xros`), `ld` notices something is wrong and fails with:
```
ld: building for 'visionOS', but linking in object file ... built for 'macOS'
```

I almost think this is a mistake I've made in the Rust side of things, but I don't know what it is. I started to log (inside of LLVM) the determined target OS, and at the point it switches to `aarch64-apple-macos` building `aarch64-apple-xros`, it starts reporting `xrOS` as the correct OS. Examining the binaries shows the MachO build command `LC_BUILD_VERSION` with the macOS id. So LLVM seems to know what we're talking about, and either is just not emitting the correct target OS (and potentially also the right code), or `rustc` is somehow instructing LLVM incorrectly.

I'm not really sure how to test LLVM in isolation to see if I can produce a suitable working binary. Going through `clang` would be too much of a pain, I think.

**Comment by agg23** [comment]
> FYI, I don't think the `x86_64-apple-xros-simulator` target will be valid

Yes, I was silently ignoring this, but I don't believe you can run the simulator on anything besides Apple Silicon.

**Comment by agg23** [comment]
Apple has closed the issue, claiming that the LLVM support is complete (it's definitely not _complete, but it may be functional). This suggests that I'm doing something wrong to prevent `rustc` from properly building the binaries.

**Comment by Nilstrieb** [comment]
(btw, I recommend opening a new zulip stream and posting updates their instead of here, which sends out GitHub notifications to many people)

**Comment by agg23** [comment]
I got things working, though it's a complicated setup. Hopefully there's not a problem with me posting it here as a discoverable way for people to find it.

My forks are:
https://github.com/agg23/rust/tree/xrOS-Target2
https://github.com/agg23/libc
https://github.com/agg23/llvm-project/tree/xrOS-Apple

# Simulator Instructions

These should work for the actual target as well; just replace the `-sim` values.

1. [Install `xargo`](https://github.com/japaric/xargo)
2. Clone https://github.com/agg23/rust.git and https://github.com/agg23/libc.git to the same root directory
3. Enter the `rust` directory. `Cargo.toml` already contains a `libc` patch to reference our cloned repo. Add this `config.toml`:
```
change-id = 102579
# These defaults are meant for users and distro maintainers building from source, without intending to make multiple changes.
[build]
# When compiling from source, you almost always want a full stage 2 build,
# which has all the latest optimizations from nightly.
build-stage = 2
test-stage = 2
doc-stage = 2
# When compiling from source, you usually want all tools.
extended = true

target = ["aarch64-apple-darwin"]

# Most users installing from source want to build all parts of the project from source.
[llvm]
download-ci-llvm = false
assertions = true
release-debuginfo = true
[rust]
# We have several defaults in bootstrap that depend on whether the channel is `dev` (e.g. `omit-git-hash` and `download-ci-llvm`).
# Make sure they don't get set when installing from source.
channel = "nightly"
download-rustc = false
deny-warnings = false

[dist]
# Use better compression when preparing tarballs.
compression-profile = "balanced"
```
4. Run `cargo update -p libc` to add the `libc` patch to `Cargo.lock` and ensure it's picked up
5. Run `./x.py build --stage 2` to build the compiler. We need to build the library at stage 2 for the _host_, not the target
6. Link toolchain via `rustup toolchain link xrOS-toolchain build/host/stage2`. `xrOS-toolchain` can be named whatever you want
7. Add `rust-toolchain.toml` and `Xargo.toml` files to your project

`rust-toolchain.toml`:
```
[toolchain]
channel = "xrOS-toolchain"
```

`Xargo.toml`:
```
[patch.crates-io]
libc = { git = "https://github.com/agg23/libc.git", rev = "db0fbbf6b554915eaffd42fb8703ab0b11625573" }

[target.aarch64-apple-xros-sim.dependencies.std]
features = ["panic-unwind"]
```
8. Force using the custom `libc` with: `xargo update -p libc --precise 0.2.151` (updates `Cargo.lock`)
9. Build the project with `xargo build --target aarch64-apple-xros-sim`

# Questions

If you have any questions or suggestions that are outside of the work required to integrate this target into Rust, [please post on the Zulip chat](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Add.20support.20for.20visionOS.20targets.20-.20rust.2Fcompiler-team.23642).

**Comment by coolbluewater** [comment]
@agg23, thanks! I tried this and looks like there are multiple issues:
1) The instructions don't mention which branch of the "rust" repo should be used. (Guess: xrOS-Target2)
2) The entry for the 'llvm-project' submodule in .gitmodules does not point to your fork.

**Comment by agg23** [comment]
Thank you. Fixed and instructions updated.

I would provide toolchain binaries so people can avoid compiling Rust, but it turns out the built toolchain is over 1GB, so I don't think it's worth the effort.

**Comment by coolbluewater** [comment]
Thanks, I updated my earlier comment, please check back there and edit your own comment to reply.

**Comment by agg23** [comment]
Currently waiting for https://github.com/rust-lang/rust/pull/120055 to merge. PRs have been submitted to `object` and `libc` in preparation for the target.

**Comment by eugenehp** [comment]
@agg23 looks like https://github.com/rust-lang/rust/pull/120055 is already merged.

**Comment by agg23** [comment]
The tier 3 target is now merged into the Rust compiler.

**Comment by eugenehp** [comment]
Amazing, thanks for pushing this! @agg23

**Comment by asdfugil** [comment]
> FYI, I don't think the `x86_64-apple-xros-simulator` target will be valid because it's called out very clearly in the Xcode 15.2 [release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_2-release-notes):
> 
> > Developing for visionOS requires a Mac with Apple silicon. (114799042)

It is still valid because the Xcode cloud servers still runs on x86, and the visionOS simulator has to work there somehow for UI tests:

You can test locally on intel with:
defaults write com.apple.dt.Xcode AllowUnsupportedVisionOSHost -bool YES
defaults write com.apple.CoreSimulator AllowUnsupportedVisionOSHost -bool YES
xcodebuild -downloadPlatform visionOS

**Comment by madsmtm** [comment]
Thanks for the heads-up @asdfugil! I'm not sure of the procedure for getting this target added, but I'd be fine with co-maintaining `x86_64-apple-visionos-sim` alongside the other visionOS targets.

**Comment by agg23** [comment]
I wonder if it needs a whole separate target policy PR like https://github.com/rust-lang/rust/pull/121419. The change should be very simple.

**Comment by workingjubilee** [comment]
@agg23 formally, yes.

I think saying "all my answers are the same as #121419 given this is basically the same target, except as follows:" should be fine though.
